container/kvm/cache.go

```diff
@@ -0,0 +1,362 @@
+	SHA256       string `yaml:"sha256,omitempty"`
+	Size         int64  `yaml:"size,omitempty"`
+	fileHandle   *os.File
+	mdHandle     *os.File
+}
+
+// Type returns the type of container this image is for. This is required by
+// for imagestorage.Metadata in the imagestorage.Storage.AddImage method call.
+func (i *Image) Type() string {
+	return "kmv"
+}
+
+// Close closes the file handles embedded in Image.
+func (i *Image) Close() error {
+	err := i.mdHandle.Close()
+	if err != nil {
+		err = errors.Trace(err)
+	}
+	err2 := i.fileHandle.Close()
+	if err2 != nil {
+		err = errors.Trace(err2)
+	}
+	return err
+}
+
+// write saves the stream to disk and updates the metadata file as well.
+func (i *Image) write(r io.Reader, md *imagedownloads.Metadata) error {
+	// First empty the existing file.
+	err := i.fileHandle.Truncate(0)
+	if err != nil {
+		return errors.Errorf("got %s when truncating %s",
+			err.Error(),
+			i.FilePath)
+	}
+	wrote, err := io.Copy(i.fileHandle, r)
+	if err != nil {
+		defer i.cleanup()
+		return errors.Trace(err)
+	}
+	if wrote != md.Size {
+		// This should likely be an error, with cleanup called prior to
+		// returning.
+		logger.Errorf("wrote %d, but expected to write %d", wrote, md.Size)
+	}
+	// We set ModTime to the time we got the update, now.
+	i.ModTime = time.Now().Format(http.TimeFormat)
+	// Size and hash are the only members that need updating.
+	i.Size = md.Size
+	i.SHA256 = md.SHA256
+	// TODO(ro) 2016-11-15 validate size and hash.
+	b, err := yaml.Marshal(i)
+	if err != nil {
+		defer i.cleanup()
+		return errors.Trace(err)
+	}
+	err = i.mdHandle.Truncate(0)
+	_, err = i.mdHandle.Write(b)
+	if err != nil {
+		defer i.cleanup()
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// cleanup attempts to remove the imagefile and metadata. It can be called if
+// things don't work out. E.g. sha256 mismatch, incorrect size...
+func (i *Image) cleanup() {
+	err := os.Remove(i.FilePath)
+	if err != nil {
+		logger.Errorf("got error %q removing %q", err.Error(), i.FilePath)
+	}
+	err = os.Remove(i.MetadataPath)
+	if err != nil {
+		logger.Errorf("got %q removing %q", err.Error(), i.MetadataPath)
+	}
+
+}
+
+// New returns a new Image from a simplestreams derived url and a pathfinder
+// func. pathfinder is the signature provided by juju.paths.DataDir. If the
+// image is already cached the returned image is populated with the cached data. If
+// not, it populates it with the metadata passed in.
+func maybeNewImage(md *imagedownloads.Metadata, pathfinder func(string) (string, error)) (*Image, error) {
+
+	// Setup names and paths.
+	dlURL, err := md.DownloadURL()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	fname := path.Base(dlURL.String())
+	baseDir, err := pathfinder(series.HostSeries())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err = maybeMakeDirs(baseDir); err != nil {
+		return nil, errors.Trace(err)
+	}
+	fpath := baseDir + "/kvm/images/" + fname
```

Comment:
Changed so there's only one directory and we don't set it up here anymore.  They would have been worked fine on windows since the base path is UNC (C:/Foo/bar). However, if that ever changed it would break subtly. So changed to using filepath join.